### PR TITLE
Feat/eng 393 provider time round

### PR DESCRIPTION
**Pull request by dardelet**
https://transmute.atlassian.net/browse/ENG-393

### CODE REVIEW

**Comment by dardelet** [comment]
WIP, do not merge

File: contracts/RoundManager.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  using SafeMath for uint;
+
+  // Round number of the last round
+  uint public lastRound;
+  // Block number of the start of the last round
+  // N.B: not necessarily the block at which initializeRound() was called
+  uint public startOfLastRound;
+
+  uint public roundLength;
+  uint public rateLockDealine;
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: Typo - rateLockDeadline*

File: contracts/RoundManager.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+
+contract RoundManager {
+  using SafeMath for uint;
+
+  // Round number of the last round
+  uint public lastRound;
+  // Block number of the start of the last round
+  // N.B: not necessarily the block at which initializeRound() was called
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: What does N.B stand for? I haven't seen this before.

File: contracts/RoundManager.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  uint public startOfLastRound;
+
+  uint public roundLength;
+  uint public rateLockDealine;
+
+  constructor() {
+    roundLength = 20;
+    rateLockDealine = 5;
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: Fix typo.

File: contracts/RoundManager.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  function timeSinceBeginningOfLastRound() internal view returns (uint) {
+    return block.number.sub(startOfLastRound);
+  }
+
+  modifier onlyBeforeActiveRoundIsLocked() {
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: Modifiers should be placed before the constructor.

File: contracts/RoundManager.sol [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    require(timeSinceBeginningOfLastRound() < roundLength.sub(rateLockDealine));
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: Fix typo.

File: contracts/RoundManager.sol [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    _;
+  }
+
+  function initializeRound() external {
```
Comment (eolszewski) [inline contracts/RoundManager.sol]: Unless this is called every block or *exactly* every `roundLength`, you will have loss of precision with rounds.

Ex: Stacktrace
```
roundLength = 5
lastRound = 0
```

initializeRound called on block 6 
```
blockNumber = 5
currentRound = 1
lastRound = 1
startOfLastRound = 5
```

initializeRound called on block 23
```
blockNumber = 23
currentRound = 4
lastRound = 4
startOfLastRound = 20
```

The round did not start on block 20, it started on 23. This effectively nullifies `roundLength` and leads to rounds of indeterminate length.

File: contracts/RoundManager.sol (hunk 4 above)
Comment (dardelet) [reply contracts/RoundManager.sol]: Thanks, didn't know. Who recommended that ? It's not in the style guide

File: contracts/RoundManager.sol (hunk 6 above)
Comment (dardelet) [reply contracts/RoundManager.sol]: You are correct. I took inspiration from Livepeer's implementation, and wondered about the same thing. I will ask them directly about this particular choice

File: contracts/RoundManager.sol (hunk 4 above)
Comment (eolszewski) [reply contracts/RoundManager.sol]: It's just a commonly accepted practice by the community.

File: test/TestRoundManager.spec.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      rm = await RoundManager.deployed();
+      electionPeriodLength = await rm.electionPeriodLength.call();
+    });
+
+    beforeEach(async () => {
+      // Adds block so that the next block is the beginning of a next round 
+      // ie adds blocks so that (blockNumber + 1) % electionPeriodLength === 0
+      await blockMiner.mineUntilEndOfElectionPeriod(rm);
+      assert.equal(0, (web3.eth.blockNumber + 1) % electionPeriodLength);
+      await rm.initializeRound();
```
[the remark this answers is not in the document]
Comment (eolszewski) [reply test/TestRoundManager.spec.js]: Agree re: happy and error cases.

Disagree: separate tests - all that is being tested here is round initialization and the code is not very complex.

**Comment by dardelet** [comment]
@eolszewski @OR13 I'm merging this because I need to submit another PR. Feel free to revert if you have any additional comments

**Pull request merged**